haystack/utils/type_serialization.py

```diff
@@ -91,60 +104,62 @@ def deserialize_type(type_str: str) -> Any:
         frozenset: typing.FrozenSet,
     }
 
-    def parse_generic_args(args_str):
-        args = []
-        bracket_count = 0
-        current_arg = ""
-
-        for char in args_str:
-            if char == "[":
-                bracket_count += 1
-            elif char == "]":
-                bracket_count -= 1
-
-            if char == "," and bracket_count == 0:
-                args.append(current_arg.strip())
-                current_arg = ""
-            else:
-                current_arg += char
-
-        if current_arg:
-            args.append(current_arg.strip())
-
-        return args
-
+    # Handle generics
     if "[" in type_str and type_str.endswith("]"):
-        # Handle generics
         main_type_str, generics_str = type_str.split("[", 1)
         generics_str = generics_str[:-1]
 
         main_type = deserialize_type(main_type_str)
-        generic_args = tuple(deserialize_type(arg) for arg in parse_generic_args(generics_str))
+        generic_args = [deserialize_type(arg) for arg in _parse_generic_args(generics_str)]
+        generic_args = tuple(generic_args) if len(generic_args) > 1 else generic_args[0]  # type: ignore
```

Comment:
Why `type: ignore`?
I would like to avoid it if possible or, otherwise, specify in the comment why it's needed.

(Same below)